docs/src/references/agent/operation-workflow.md

```diff
@@ -236,6 +236,19 @@ Data extracted from the command status topic and message payload can be passed a
   - `${.payload.status}` is the command current status (*e.g.* `"init"`)
   - `${.payload.x.y.z}` is the json value extracted from the payload following the given `x.y.z` path if any.
   - If given `${.some.unknown.path}`, the argument is passed unchanged to the script.
+  - An argument can contain a combination of path expressions and literal:
+    - in a context where the operation instance payload is  `{"x":"X", "y":"Y"}`
+    - the template `"prefix-${.payload.x}-separator-${.payload.y}-suffix"`
+    - is replaced by `"prefix-X-separator-Y-suffix"`.
+- These substitutions are applied using the following rules:
+  - The script command is first tokenized using shell escaping rules.
+    *e.g.* `"sudo systemctl restart tedge-mapper-${.payload.mapperType}"`
+    is interpreted as the `sudo` command with 3 parameters.
+  - Then all the path expressions of each parameter are expended using the command current topic and payload.
+    *e.g.* `"tedge-mapper-${.payload.mapperType}"` is interpreted as `"tedge-mapper-c8y"`
+    (assuming the JSON payload contains a `mapperType` field with `"c8y"` as string value).
+  - These substitutions apply to the command itself as well as all its arguments.
+  - Any ill-formed path expression is let unchanged.
```

Comment:
```suggestion
  - Any ill-formed path expression is left unchanged.
```

Reply:
Though do we class non-existent paths as "ill-formed"?

For example, if a non-existent payload property is referenced, then it makes it more difficult for the binary to determine if the value is an expanded or unexpanded value:

```toml
script = "/usr/bin/mybinary ${.payload.doesNotExist}"
```

Checking for unexpanded values in a bash script is not so trivial, it would involve the following regex comparison:

```sh
#!/usr/bin/env bash

# check if value is expanded
if [[ "$1" =~ ^\\$\{.+\}$ ]]; then
    echo "Value is not expanded" >&2
    exit 1
fi
```

However if we just expanded the value to an empty string, then it becomes easier to detect (and generally a script does not care about the template variable language, just the value at the end)

```sh
#!/usr/bin/env bash

# check if value is empty
if [ -z "$1" ]; then
    echo "Value is empty" >&2
    exit 1
fi
```

Reply:
The distinction between what is ill-formed and what is non-existent is more applicable to the `${.payload}` object. Since the payload object represents a free-form value, then any nested paths should be allowed (regardless if the nested path exists under the payload or not).

```sh
# ok
${.payload.<path>}

# not ok
${.payld.<path>}
${.tpc}
```

Reply:
Okay. I will change the behavior for unknown path (as in `${.payload.doesNotExist}`) to replace the path expression by an empty string. The intent of letting the expression unchanged in such a case was to help debugging by stressing something unexpected. I agree there are pros and cons and that an empty string can be more convenient in practice. Notably, if the user want to use message payloads with unset optional fields.

However, making a difference between `${.payload.<unknown-path>}` and `${.unknow-root.<path>}` might be more subtle. I will see what can be done.

Reply:
> The distinction between what is ill-formed and what is non-existent is more applicable to the `${.payload}` object. Since the payload object represents a free-form value, then any nested paths should be allowed (regardless if the nested path exists under the payload or not).
> 
> ```shell
> # ok
> ${.payload.<path>}
> 
> # not ok
> ${.payld.<path>}
> ${.tpc}
> ```

This has been addressed by https://github.com/thin-edge/thin-edge.io/pull/2918/commits/ccd8d2677f297b978c6ef075cfa242d1ab099cda

Reply:
great work